routes/signMessage.ts

```diff
@@ -1,21 +1,104 @@
-import { signMessage } from '../lib/signatures'
-import { app } from '../server'
+import { ed25519 } from '@noble/curves/ed25519'
+import { blake3 } from '@noble/hashes/blake3'
+import { app } from '../app'
+import { messageDataToUint8Array } from '../buffers/buffers'
+import { kms } from '../clients/aws'
+import { authDb } from '../database/watcher'
+import { selectKeysQuery, selectSessionQuery } from '../lib/queries'
+import { signWithEddsaKey } from '../utils/signatures'
+import { lucia } from '../lucia/auth'
+import { base16, base64 } from '@scure/base'
 
-app.post('/sign', async (c) => {
+import type { Message } from '../utils/types'
+import { isMessage } from '../utils/types'
+
+app.post('/signMessages', async (c) => {
   try {
-    const { message } = await c.req.json()
+    const { sessionId, messages } = await c.req.json()
+
+    if (!sessionId || !messages || !Array.isArray(messages)) {
+      return c.json(
+        { success: false, message: 'Missing or invalid parameters' },
+        400,
+      )
+    }
+
+    for (const message of messages) {
+      if (!isMessage(message)) {
+        return c.json(
+          { success: false, message: 'Invalid message format' },
+          400,
+        )
+      }
+    }
+
+    const { session } = await lucia.validateSession(sessionId)
+
+    if (!session) {
+      return c.json({ success: false, message: 'Invalid session' }, 404)
+    }
+
+    const sessionResult = await authDb.query(selectSessionQuery, [sessionId])
+    const { userid } = sessionResult.rows[0]
+
+    const keysResult = await authDb.query(selectKeysQuery, [userid])
+
+    if (keysResult.rows.length === 0) {
+      return c.json({ success: false, message: 'Keys not found' }, 404)
+    }
+
+    const { encryptedprivatekey, publickey } = keysResult.rows[0]
+
+    // Decrypt the private key using AWS KMS
+    const decryptedPrivateKey = await kms
+      .decrypt({
+        CiphertextBlob: base64.decode(encryptedprivatekey),
+      })
+      .promise()
 
-    if (!message) {
-      return c.json({ success: false, message: 'No message provided' }, 400)
+    if (!decryptedPrivateKey.Plaintext) {
+      throw new Error('Decryption failed')
     }
 
-    const signedMessage = signMessage(message)
+    const eddsaPrivateKey = new Uint8Array(
+      decryptedPrivateKey.Plaintext as ArrayBuffer,
+    )
+    const signedMessages: Message[] = []
+
+    for (const message of messages) {
+      if (userid.toString() !== message.messageData.rid) {
+        return c.json({ success: false, message: 'RID mismatch' }, 400)
+      }
+
+      const computedHash = blake3(messageDataToUint8Array(message.messageData))
+
+      if (computedHash !== message.hash) {
+        return c.json({ success: false, message: 'Invalid message hash' }, 400)
+      }
+
+      const signature = signWithEddsaKey(message.hash, eddsaPrivateKey)
+
+      const signerUInt8Array = ed25519.getPublicKey(eddsaPrivateKey)
+      const signer = base16.encode(signerUInt8Array)
```

Comment:
plz use viem for this, bc we want this to be ethereum compatibile (they have a bytesToHex func)